service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/ActionCaseManagementOrderController.java

```diff
@@ -0,0 +1,125 @@
+                                               ObjectMapper mapper,
+                                               CMODocmosisTemplateDataGenerationService templateDataGenerationService) {
+        this.draftCMOService = draftCMOService;
+        this.caseManagementOrderService = caseManagementOrderService;
+        this.docmosisDocumentGeneratorService = docmosisDocumentGeneratorService;
+        this.uploadDocumentService = uploadDocumentService;
+        this.mapper = mapper;
+        this.templateDataGenerationService = templateDataGenerationService;
+    }
+
+    @PostMapping("/about-to-start")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        Map<String, Object> data = caseDetails.getData();
+        final CaseData caseData = mapper.convertValue(data, CaseData.class);
+
+        caseDetails.getData()
+            .putAll(caseManagementOrderService.extractMapFieldsFromCaseManagementOrder(caseData.getCmoToAction()));
+
+        draftCMOService.prepareCustomDirections(caseDetails, caseData.getCmoToAction());
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(
+        @RequestHeader(value = "authorization") String authorization,
+        @RequestHeader(value = "user-id") String userId,
+        @RequestBody CallbackRequest callbackRequest) throws IOException {
+
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        Document document = getDocument(authorization, userId, caseData, false);
+
+        caseDetails.getData().put("orderAction", ImmutableMap.of("document", buildFromDocument(document)));
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(
+        @RequestHeader(value = "authorization") String authorization,
+        @RequestHeader(value = "user-id") String userId,
+        @RequestBody CallbackRequest callbackRequest) throws IOException {
+
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+        CaseManagementOrder order = caseData.getCmoToAction();
+
+        order = draftCMOService.prepareCMO(caseData, order);
+
+        order = caseManagementOrderService.addAction(order, caseData.getOrderAction());
+
+        Document document = getDocument(authorization, userId, caseData, order.isApprovedByJudge());
+
+        order = caseManagementOrderService.addDocument(order, document);
```

Comment:
This place has got a lot of logic how this build, probably could be refactored at some point into some service? Looks like this orchestrates how to build order which is not responsibility of hte controller 👺 

Do we need to generate new document here? As we generated another one in mid-event? Ideally we could generate only if changed (again not change for this PR).

Reply:
sure, I think this will probably fall into a refactor I want to do for a CaseManagementOrderService. This service will handle the building of the cmo. I don't want to touch draftCMO too much in this PR.

Reply:
You're also correct we should only generate document when change, however a change could be as simple as a character change in one question of a complex type. That is a lot of checks...